[analyzer] Workaround for slowdown spikes (unintended scope increase)

NagyDonat opened the pull request:
Recently some users reported that they observed large increases of runtime (up to +600% on some translation units) when they upgraded to a more recent (slightly patched, internal) clang version. Bisection revealed that the bulk of this increase was probably caused by my earlier commit bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 ("Don't assume third iteration in loops").

As I evaluated that earlier commit on several open source project, it turns out that on average it's runtime-neutral (or slightly helpful: it reduced the total analysis time by 1.5%) but it can cause runtime spikes on some code: in particular it more than doubled the time to analyze `tmux` (one of the smaller test projects). 

Further profiling and investigation proved that these spikes were caused by an _increase of analysis scope_ because there was an heuristic that placed functions on a "don't inline this" blacklist if they reached the `-analyzer-max-loop` limit (anywhere, on any one execution path) -- which became significantly rarer when my commit ensured the analyzer no longer "just assumes" four iterations. (With more inlining significantly more entry points use up their allocated budgets, which leads to the increased runtime.)

I feel that this heuristic for the "don't inline" blacklist is unjustified and arbitrary, because reaching the "retry without inlining" limit on one path does not imply that inlining the function won't be valuable on other paths -- so I hope that we can eventually replace it with more "natural" limits of the analysis scope.

However, the runtime increases are annoying for the users whose project is affected, so I created this quick workaround commit that approximates the "don't inline" blacklist effects of ambiguous loops (where the analyzer doesn't understand the loop condition) without fully reverting the "Don't assume third iteration" commit (to avoid reintroducing the false positives that were eliminated by it).

Investigating this issue was a team effort: I'm grateful to Endre Fülöp (gamesh411) who did the bisection and shared his time measurement setup, and Gábor Tóthvári (tigbr) who helped me in profiling.

NagyDonat committed:
[analyzer] Workaround for unintended slowdown (scope increase)

Recently some users reported that the runtime of the static analyzer
drastically increased on their codebase when they upgraded to a more
recent (sligthly patched, internal) clang version: some of their
translation units got more than +600% runtime (yes, with two zeroes) and
IIUC the average increase was also high.

Bisection revealed that the bulk of the increase was caused by my
earlier commit bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 ("Don't assume
third iteration in loops") and I verified that the slowdown affects the
downstream clang. (I used tmux as an arbitrary easy-to-analyze open
source project and measured that reverting this commit reduces the
total runtime by more than 50% compared to a recent main revision.)

Further profiling and investigation proved that this increase is
caused by an _increase of analysis scope_ because there was an arbitrary
heuristic that placed functions on a "don't inline this" blacklist if
they reached the `-analyzer-max-loop` limit (anywhere, on any one
execution path) -- which became significantly rarer when my commit
ensured the analyzer no longer "just assumes" four iterations. (With
more inlining significantly more entry points use up their allocated
budgets, which leads to the increased runtime.)

I feel that this heuristic for the "don't inline" blacklist is
unjustified, because reaching the "retry without inlining" limit on one
path does not imply that inlining the function won't be valuable on
other paths -- so I hope that we can eventually replace it with more
"natural" limits of the analysis scope.

However, the brutal jump of analysis runtime significantly degrades the
user experience, so I created this quick workaround commit that
approximates the "don't inline" blacklist effects of opaque loops (where
the analyzer doesn't understand the loop condition) without fully
reverting the "Don't assume third iteration" commit (to avoid
reintroducing the false positives that were eliminated by it).

NagyDonat commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/FunctionSummary.h:
```diff
@@ -81,10 +81,6 @@ class FunctionSummariesTy {
     I->second.MayInline = 0;
   }
 
-  void markReachedMaxBlockCount(const Decl *D) {
-    markShouldNotInline(D);
-  }
```
I'm removing this function (by inlining its trivial definition) because I feel that its nondescript sneaky name significantly contributed to the fact that I did not know about this heuristic previously.

**Please name functions clearly** – refer to their effects instead of just naming their call site (which is not helpful when I'm already reading the code that calls them)!

balazs-benics-sonarsource replied:
I'd not mind keeping this if there was more thing to do once a "Max block count" is reached. But in this particular case I agree with you.

NagyDonat replied:
I agree that it's OK to write code like
```c++
if (<reached condition X>)
  markReachedConditionX(Args);
```
where every effect is grouped into a single function, I just say that it's a bad pattern to hide a function with this kind of generic name in a large block of mixed effects:
```c++
if (<reached condition X>) {
  doSomething();
  markReachedConditionX(Args);
  if (InnerCondition) {
    makeFoo();
    return conjureBar();
  } else {
    //...
  }
}
```

NagyDonat commented on clang/test/Analysis/loop-unrolling.cpp:
```diff
@@ -1,5 +1,5 @@
-// RUN: %clang_analyze_cc1 -analyzer-checker=core,debug.ExprInspection -analyzer-config unroll-loops=true,cfg-loopexit=true -verify -std=c++14 -analyzer-config exploration_strategy=unexplored_first_queue %s
-// RUN: %clang_analyze_cc1 -analyzer-checker=core,debug.ExprInspection -analyzer-config unroll-loops=true,cfg-loopexit=true,exploration_strategy=dfs -verify -std=c++14 -DDFS=1 %s
+// RUN: %clang_analyze_cc1 -analyzer-checker=core,debug.ExprInspection -analyzer-config unroll-loops=true,cfg-loopexit=true -verify=expected,default -std=c++14 -analyzer-config exploration_strategy=unexplored_first_queue %s
+// RUN: %clang_analyze_cc1 -analyzer-checker=core,debug.ExprInspection -analyzer-config unroll-loops=true,cfg-loopexit=true,exploration_strategy=dfs -verify=expected,dfs -std=c++14 %s
```
I'm leaning towards eliminating the alternative RUN: line that uses `dfs` because the only difference between `dfs` and the default exploration algorithm is a "control group" testcase where `unroll-loops` does not activate.

I'll probably do so in a separate follow-up commit.

balazs-benics-sonarsource replied:
As you wish.

balazs-benics-sonarsource commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2523,6 +2523,20 @@ bool ExprEngine::replayWithoutInlining(ExplodedNode *N,
+    return nullptr;
+
+  return CalleeLC;
+}
```
~~Is this basically the getStackFrame's parent's getStackFrame?~~
No, you are right. There may be some other location context above the Stack frame context of the parent frame.

balazs-benics-sonarsource commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2856,8 +2873,29 @@ void ExprEngine::processBranch(
-      if (!SkipTrueBranch || AMgr.options.ShouldWidenLoops)
+      if (!SkipTrueBranch || AMgr.options.ShouldWidenLoops) {
         Builder.generateNode(StTrue, true, PredN);
+      } else if (AMgr.options.LegacyInliningPrevention) {
+        // FIXME: There is an ancient and very arbitrary heuristic in
+        // `ExprEngine::processCFGBlockEntrance` which prevents all further
+        // inlining of a function if it finds an execution path within that
+        // function which reaches the `MaxBlockVisitOnPath` limit (a/k/a
+        // `analyzer-max-loop`, by default four iterations in a loop). Adding
+        // this "don't assume third iteration" logic significantly increased
+        // the analysis runtime on some inputs because less functions were
+        // arbitrarily excluded from being inlined, so more entrypoints used
+        // up their full allocated budget. As a hacky compensation for this,
+        // here we apply the "should not inline" mark in cases when the loop
+        // could potentially reach the `MaxBlockVisitOnPath` limit without the
+        // "don't assume third iteration" logic. This slightly overcompensates
+        // (activates if the third iteration can be entered, and will not
+        // recognize cases where the fourth iteration would't be completed), but
+        // should be good enough for practical purposes.
+        if (const LocationContext *LC = getInlinedLocationContext(Pred, G)) {
+          Engine.FunctionSummaries->markShouldNotInline(
+              LC->getStackFrame()->getDecl());
+        }
+      }
```
How about if we would apply this chunk unconditionally? I don't think anyone would want to regress 6x voluntarily.

balazs-benics-sonarsource replied:
My point is, if Ericsson will ship this `true`, and we are also going to ship this with `true`, then who would want to disable this? Do you expect users disabling this? Would anyone complain if we would just land this fix without an option for opting out?
Speaking of arbitrary cut heuristics, maybe they are arbitrary, but that should only make use more careful of changing them because continuity itself is valuable - but I'm not opposed, just cautions.

NagyDonat replied:
> My point is, if Ericsson will ship this true, and we are also going to ship this with true, then who would want to disable this? Do you expect users disabling this?

Good point.

>  Would anyone complain if we would just land this fix without an option for opting out?

I felt that the logic introduced by this commit is so ugly that it should not be unconditionally part of the analyzer behavior (especially if it is `git blame`d to me...). By introducing this option I wanted to express that "this is not a stable permanent part of the analyzer, please don't rely on its existence" -- and I felt that yet another "FIXME" comment is not enough for this hack.

However, I can accept removing this option (and making the hack unconditional) if you strongly prefer that.

> Also I am not sure we should have legacy in the name of something that is enabled by default.

When I wrote this change, I hoped that this config option could be switched off and deprecated in the foreseeable future (in fact it was disabled by default in the first draft of the commit).

balazs-benics-sonarsource commented on clang/test/Analysis/legacy-inlining-prevention.c:
```diff
@@ -0,0 +1,198 @@
+  int i;
+  clang_analyzer_numTimesReached(); // expected-warning {{1}}
+  for (i = 0; i < 10; i++);
+  clang_analyzer_numTimesReached(); // no-warning
```
```suggestion
  clang_analyzer_numTimesReached(); // FIXME: It should be reachable
```

NagyDonat replied:
I would be careful with this FIXME -- if somebody follows it and fixes this issue (e.g. by enabling `unroll-loops` by default) we could get yet another large slowdown when less loops hit the `analyzer-max-loop` limit, less functions end on the "don't inline" list and more entry points spend their budget.

Obviously in an ideal world an empty loop like this shouldn't stop the analyzer – but we must replace this awkward loop-based inlining restriction with a more robust heuristic before we can do that.

balazs-benics-sonarsource replied:
I didn't get the argument why would be a risk of having a FIXME here. It's the reviewer's and maintainer's responsibility to ensure the changes align with the long term directions of the project.
If someone would just override `analyzer-max-loop` for this test "to fix the FIXME", that should be rejected.
However, if we found a way to enable it by default, or have a smarter engine, then it's fine - but that would also touch files beyond this single test file.

balazs-benics-sonarsource replied:
Sounds good.

balazs-benics-sonarsource commented on clang/test/Analysis/legacy-inlining-prevention.c:
```diff
@@ -0,0 +1,198 @@
+//-----------------------------------------------------------------------------
+// Inlined function always reaches `analyzer-max-loop`.
+
+int inner_fixed_loop_1(void) {
+  int i;
+  clang_analyzer_numTimesReached(); // expected-warning {{1}}
```
You could pass a `const char*` from a string literal to differentiate which is inlined when observing its value using a `clang_analyzer_dump()` at the beginning of this inlined function. That should make it clear that for the first time its inlined, but for the second time it's not.

balazs-benics-sonarsource commented on clang/test/Analysis/legacy-inlining-prevention.c:
```diff
@@ -0,0 +1,198 @@
+}
+
+int outer_fixed_loop_1(void) {
+  int x = inner_fixed_loop_1();
+  int y = inner_fixed_loop_1();
```
We should have comments here that first it's inlined, second it's not because of the given heuristic.

balazs-benics-sonarsource commented on clang/test/Analysis/legacy-inlining-prevention.c:
```diff
@@ -0,0 +1,198 @@
+int outer_fixed_loop_1(void) {
+  int x = inner_fixed_loop_1();
+  int y = inner_fixed_loop_1();
+  return 53 / (x - y); // no-warning
```
My problem with these `no-warnings` in general in this PR that it documents what the test currently does, but what they should document what the tests should/could expect.
In this case in an ideal world, we should actually get a diagnostic, thus the desired outcome is not a `no-warning`.
Consequently, a FIXME would be more appropriate I think.

balazs-benics-sonarsource commented on clang/test/Analysis/legacy-inlining-prevention.c:
```diff
@@ -0,0 +1,198 @@
+  int i;
+  clang_analyzer_dump(arg); // expected-warning {{2}}
+  for (i = 0; i < 10; i++);
+  clang_analyzer_dump(arg); // no-warning
```
FIXME.

balazs-benics-sonarsource reviewed (changes requested):
Thank you for investigating this. At Sonar, we have not yet started the upgrade to clang-20. I suppose, you already did then, thus found this regression on trunk.

Maybe we should also reflect of the quality control of our submissions of core changes too, but let's leave that after the PR is discussed, and we put out the fire.

My impression is that we shouldn't have a new flag for this, we should unconditionally apply this for now. Did you think about this?

I left a few other comments about testing but overall I'm all for this change.
Thanks again!

balazs-benics-sonarsource commented:
One other note. We should backport this fix to clang-20 once it lands to main.

NagyDonat committed:
Use clang_analyzer_dump in tests

Note that we need to expect e.g. {{2 S32}} because the pattern {{2}}
would match any other 32-bit integer as well.

The last two testcases in the file were testing the same thing in
different ways, so this commit only keeps the one that used
`clang_analyzer_dump` (which was previously an exception).

NagyDonat commented:
I tested the impact of this commit on various open source projects and the changes in the result counts are surprisingly small:

| Project | New Reports | Resolved Reports |
|---------|-------------|------------------|
| memcached | 0 new reports | 0 resolved reports 
| tmux | 2 new reports | 6 resolved reports 
| curl | 0 new reports | 0 resolved reports 
| twin | 3 new reports | 2 resolved reports 
| vim | 11 new reports | 18 resolved reports 
| openssl | 0 new reports | 1 resolved reports 
| sqlite | 28 new reports | 8 resolved reports 
| ffmpeg | 22 new reports | 23 resolved reports 
| postgres | 3 new reports | 20 resolved reports 
| tinyxml2 | 0 new reports | 0 resolved reports 
| libwebm | 1 new reports | 0 resolved reports 
| xerces | 3 new reports | 5 resolved reports 
| bitcoin | 0 new reports | 0 resolved reports 
| protobuf | 2 new reports | 3 resolved reports 
| qtbase | 27 new reports | 23 resolved reports 
| contour | new reports not found | resolved reports not found 
| openrct2 | 0 new reports | 7 resolved reports 

Unfortunately there was some trouble with our publicly available "demo" CodeChecker server (where we usually present the analysis results), so I had to store these results to an internal server and they are not available externally. (If somebody is interested, I can give priority to finding a way to publish the full report set.)

steakhal commented:
Do you have data about the analysis times per file, and per analysis entry point?
Compared against the current llvm main, and also if this workaround would restore the original running times before https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849?

balazs-benics-sonarsource commented:
Do you have data about the analysis times per file, and per analysis entry point?
Compared against the current llvm main, and also if this workaround would restore the original running times before https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849?

NagyDonat commented on clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:
```diff
@@ -414,6 +414,20 @@ ANALYZER_OPTION(
+    "Note that functions that _definitely_ reach the loop limit on some "
+    "execution path are currently marked as \"do not inline\" even if this "
+    "option is disabled (but this may change in future versions). This option "
+    "is a dumb and arbitrary restriction on inlining, but disabling it would "
```
Good point, I was annoyed by the situation when I wrote this description, but I was already planning to rewrite these before finalizing the commit.

NagyDonat commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2570,21 +2584,24 @@ void ExprEngine::processCFGBlockEntrance(const BlockEdge &L,
-      Engine.FunctionSummaries->markReachedMaxBlockCount(CalleeSF->getDecl());
+    if (const LocationContext *LC = getInlinedLocationContext(Pred, G)) {
+      // FIXME: This will unconditionally prevent inlining this function (even
+      // from other entrypoints), which is not a reasonable heuristic: even if
```
Thanks for spotting this! For some reason I tend to automatically write "entrypoint"; before uploading the PR I fixed it in many locations, but apparently I failed to catch this one...

NagyDonat commented:
@balazs-benics-sonarsource Unfortunately our current CI (which runs the static analyzer via CodeChecker) cannot provide file-level or entrypoint-level timing information, but I do have full analysis runtime measurements (which I'll edit into the table in my previous comment).

This comparison is between `legacy-inlining-prevention=false` (which is practically equivalent to the recent main revision a84a6f7dd68b218757e192fe21a806c80ef0b63d) and `legacy-inlining-prevention=true` (the default in this patch), but I can measure timing differences compared to other commits as well. I think the most natural comparison would be between
- this commit with its default `legacy-inlining-prevention=true` mode and
- reverting bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 ("Don't assume third iteration") atop a84a6f7dd68b218757e192fe21a806c80ef0b63d (the parent of this commit)
because that way the comparison wouldn't be influenced by other commits that are merged in the last few months.

NagyDonat committed:
Correct docs and comments

NagyDonat commented:
I pushed two commits (including https://github.com/llvm/llvm-project/pull/136720/commits/d57946b9b26e9244f5d3819956ffc767d0e0b611 which is hidden among the earlier review comments based on its creation time), and these in theory address most review comments, with the exception of two areas:
- Writing a detailed FIXME note about the "analysis is interrupted if it reaches four iterations" question.
- Changing this hack from "enabled by default config" to "unconditionally enabled" -- or just renaming the config.

I'll try to address these by tomorrow.

NagyDonat commented:
Execution time comparisons between the commit under review and just reverting https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 ("Don't assume third iteration") atop https://github.com/llvm/llvm-project/commit/a84a6f7dd68b218757e192fe21a806c80ef0b63d (the parent of the commit under review):

| Project | just revert | commit under review | Ratio |
| --- | --- | --- | --- |
| memcached | 27.50 | 25.68 | 93.4% |
| tmux | 79.49 | 73.16 | 92.0% |
| curl | 133.26 | 117.82 | 88.4% |
| twin | 54.28 | 48.40 | 89.2% |
| vim | 221.86 | 165.80 | 74.7% |
| openssl | 304.98 | 256.52 | 84.1% |
| sqlite | 1039.42 | 870.52 | 83.8% |
| ffmpeg | 1272.77 | 1190.92 | 93.6% |
| postgres | 402.34 | 321.95 | 80.0% |
| tinyxml2 | 49.82 | 28.65 | 57.5% |
| libwebm | 83.14 | 56.02 | 67.4% |
| xerces | 202.53 | 175.92 | 86.9% |
| bitcoin | 162.05 | 155.26 | 95.8% |
| protobuf | 2367.55 | 1969.73 | 83.2% |
| qtbase | 3022.01 | 2729.55 | 90.3% |
| openrct2 | 889.14 | 782.47 | 88.0% |

:thinking: The version with the commit under review is surprisingly fast and I don't exactly know why. My most plausible theory is that https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 ("Don't assume third iteration") has two effects on the analysis runtime:
- it inherently speeds up the analysis of loops (because some iterations are skipped);
- it slows down the analysis because it affects the inlining heuristic and prevents functions from being placed on the inlining blacklist.
This would explain why just reverting that commit (which undoes both effects) produces slower analysis than applying the commit under review (which undoes the slowdown and keeps the speedup). 

One plausible explanation is that https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849 "Don't assume third iteration" implies an inherent speedup, because it prevents

NagyDonat committed:
Rename the new test file

NagyDonat committed:
Document loop handling limitations that are required by the inlining prevention

NagyDonat commented:
> But I was shocked that sometimes intuition fails, and we didn't check the RT for that patch.

This is a legacy project with many skeletons hidden in various closets... The main moral of this story is probably that we should be more paranoid about measuring runtime (I'm thinking about whether we can set up some regular or semi-regular runtime measurements in our CI), but this also shows why we should work on reducing the technical debt.

NagyDonat commented:
With the two recent commits I think I addressed all review comments except for the "on-by-default config option" vs "unconditionally active logic" question.

In this area (which we discussed in inline comments) I'm still leaning towards the currently implemented `legacy-inlining-prevention` config option (which is enabled by default), because this would provide very minor advantages in several areas:
- The average slowdown caused by disabling this workaround seems to be around 10-15% on larger projects, which is painful, but not completely prohibitive (and I guess that it could be offset if the user e.g. reduces `max-nodes` by 20%).
- If somebody happens to run into false negatives (or invalidation-based false positives) which are caused by this inlining prevention, we can suggest `legacy-inlining-prevention=false` as a workaround.
- I hope that I will be able to develop other heuristics that prevent the inlining of complex functions based on more "natural" and robust conditions -- when those are available, some users may want to disable this and enable those instead. (Or, if they are proven to be better and we change the defaults, this option should remain as a disabled-by-default fallback for those who might want return to the earlier behavior.)
- During the development and testing of those alternative heuristics it would be nice to have this config option to enable/disable this workaround.
- Last, but perhaps not least, having a named config option will make it easier to speak about this code fragment (instead of "that hack where we disable inlining of a function if we would assume entering the third iteration in a loop").

If we keep the config option, we can also think about renaming it. I agree that `legacy` is a bit unfortunate in the name of an option that is on-by-default (at least for now), but I don't have a better idea. (Feel free to suggest names if you have!)

@balazs-benics-sonarsource @Xazax-hun WDYT about this area? Do you have any other question/remark?

balazs-benics-sonarsource commented:
I can see your points. I think they indeed moves the needle slightly for having this (or a similar) flag but barely.
I'd need to think about a better flag name, but right now I'm too busy for that.

NagyDonat commented:
I did yet another measurement where I measured the runtime of [bb27d5e](https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849) ("Don't assume third iteration") compared to its parent commit (i.e. the measurement that we should've done before merging it) and it turns out that **on average it does not increase the runtime** and tmux (where Endre observed a significant slowdown) is an extreme outlier:
| Project | without [bb27d5e](https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849) | with [bb27d5e](https://github.com/llvm/llvm-project/commit/bb27d5e5c6b194a1440b8ac4e5ace68d0ee2a849) | Ratio |
| --- | --- | --- | --- |
| memcached | 23.25 | 22.26 | 95.7% |
| tmux | 67.47 | 146.70 | 217.4% |
| curl | 116.58 | 111.74 | 95.8% |
| twin | 42.81 | 40.33 | 94.2% |
| vim | 164.54 | 154.02 | 93.6% |
| openssl | 274.13 | 258.72 | 94.4% |
| sqlite | 902.19 | 943.20 | 104.5% |
| ffmpeg | 1161.76 | 1130.22 | 97.3% |
| postgres | 367.47 | 335.01 | 91.2% |
| tinyxml2 | 36.11 | 45.79 | 126.8% |
| libwebm | 75.85 | 64.33 | 84.8% |
| xerces | 191.27 | 212.58 | 111.1% |
| bitcoin | 159.05 | 159.09 | 100.0% |
| protobuf | 2249.66 | 2108.13 | 93.7% |
| qtbase | 2796.47 | 2812.94 | 100.6% |
| openrct2 | 946.48 | 880.49 | 93.0% |
| sum of all projects | 9575.09 | 9425.55 | 98.4% |

By the way Endre Fülöp picked tmux arbitrarily as a "random small open source project that is easy to build" when he tried to reproduce the slowdown reported by the user and as it immediately reproduced a significant slowdown, we assumed that this is an universal phenomenon -- but it turns out that this was just a coincidence.

These results show that most users won't experience a brutal slowdown (especially if they have a large varied codebase where the law of large numbers guarantees average results), which reduces the urgency of this PR a bit. (And I will rewrite the commit message, clarifying that the runtime increases are mostly concentrated on individual TUs and aren't too large on average.)

NagyDonat commented:
@balazs-benics-sonarsource Gentle ping

balazs-benics-sonarsource commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2523,6 +2523,20 @@ bool ExprEngine::replayWithoutInlining(ExplodedNode *N,
+static const LocationContext *getInlinedLocationContext(ExplodedNode *Node,
+                                                        ExplodedGraph &G) {
+  const LocationContext *CalleeLC = Node->getLocation().getLocationContext();
+  const LocationContext *RootLC =
```
This wasn't addressed.

balazs-benics-sonarsource replied:
Awesome, thank you for checking!

balazs-benics-sonarsource reviewed (commented):
Looks good. There were two points unaddressed:
 - Finding a name for the flag without the `legacy-` prefix
 - Find out if we can ever have multiple root nodes in an exploded graph.

NagyDonat commented:
I discussed the name with Endre (@gamesh411) in person and our best idea is `inline-functions-with-ambiguous-loops` which would be the opposite of the currently implemented `legacy-inlining-prevention` (i.e. the default would be `inline-functions-with-ambiguous-loops=false` which is equivalent to `legacy-inlining-prevention=true`).

@balazs-benics-sonarsource @anybody else 
What do you think about this name? Would this be acceptable for you?

balazs-benics-sonarsource commented:
@Xazax-hun WDYT of the proposed alternative flag name?

Xazax-hun commented:
`inline-functions-with-ambiguous-loops` sounds good to me. I don't think it makes sense to have `legacy` in the name of an on by default setting.

NagyDonat committed:
Rename and flip the config option

balazs-benics-sonarsource reviewed (approved):
Thank you!

NagyDonat committed:
Fix "the the" typo

balazs-benics-sonarsource commented:
Thanks for the fix, I'll proceed with the backport if you still believe it's worthy. @NagyDonat

NagyDonat commented:
This is not an automatic "every project is affected" slowdown, but on smaller outlier projects like tmux the runtime may double, so I'd guess that it's reasonable to backport this commit (although I'm not familiar with the exact conditions when a commit should be backported). Thanks for doing the backport (if you decide to do it).

steakhal commented:
Backport requested in #139597
